returnn/datasets/postprocessing.py

```diff
@@ -0,0 +1,251 @@
+        return True
+
+    def _collect_single_seq(self, seq_idx: int) -> Optional[DatasetSeq]:
+        assert seq_idx > self._data_iter_seq_idx, "collecting seqs must be done strictly monotonically"
+        self._data_iter_seq_idx = seq_idx
+
+        while True:
+            try:
+                loaded_seq_idx, tensor_dict = next(self._data_iter)
+            except StopIteration:
+                return None
+            if loaded_seq_idx != seq_idx:
+                continue
+            seq = DatasetSeq(features=tensor_dict.as_raw_tensor_dict(exclude_sequence_lengths=True), seq_idx=seq_idx)
+            return seq
+
+    def _build_dataset_iter(self, dataset: Dataset) -> Iterator[TensorDict]:
+        """
+        :return: an iterator applying both the segment level and across-segment transformations on the given dataset
+        """
+
+        def _wrap_map_seq():
+            for item in self._iterate_dataset(dataset):
+                mapped_item = self._map_seq(item)
+                if isinstance(mapped_item, Iterator):
+                    yield from mapped_item
+                else:
+                    yield mapped_item
+
+        map_iter = self._map_seq_stream(_wrap_map_seq())
+        assert isinstance(map_iter, Iterator), "map_seq_stream must produce an iterator"
+        return map_iter
+
+    def _data_dict_to_tensor_dict(self, dataset: Dataset, data_dict: Dict[str, ndarray]) -> TensorDict:
+        """
+        :return: the given data dict converted to a TensorDict class
+        """
+
+        def _make_tensor(name: str, data: ndarray) -> Tensor:
+            sparse_dim = None
+            if data.dtype.name.startswith("str"):
+                dims = []
+                dtype = "string"
+            else:
+                dims, sparse_dim = self._dim_cache.get(name, (None, None))
+                dtype = data.dtype
+
+            if dims is None:
+                feature_dims = [
+                    Dim(dimension=v, name=f"{name}_dim{i + 1}") for i, v in enumerate(dataset.get_data_shape(name))
+                ]
+                dims = [Dim(dimension=None, name=f"{name}_num_frames"), *feature_dims]
+                if dataset.is_data_sparse(name):
+                    sparse_dim = Dim(
+                        dimension=dataset.get_data_dim(name) if dataset.is_data_sparse(name) else None,
+                        name=f"{name}_sparse",
+                    )
+                self._dim_cache[name] = (dims, sparse_dim)
+
+            try:
+                return Tensor(name, dims=dims, dtype=dtype, sparse_dim=sparse_dim, raw_tensor=data)
+            except Exception as exc:
+                raise Exception(
+                    "Could not convert from mapping function output to `TensorDict`, "
+                    f"do the data shapes {data.shape} match up with the ones declared "
+                    f"in `map_output` (inferred to: {dims}, sparse={sparse_dim})?"
+                ) from exc
```

Comment:
Don't do such generic all exception catching. Your code should already catch any errors, and then you should expect that there is no way that any further exception is raised (i.e. an exception here would be unexpected).
```suggestion
            return Tensor(name, dims=dims, dtype=dtype, sparse_dim=sparse_dim, raw_tensor=data)
```

Reply:
The idea is to help the user catch any dim mismatch errors more quickly that occur due to the mapping function returning data in the wrong shape. I think the issue here is that `Tensor`, during its internal shape validation, just raises a generic `Exception`, and not a more specific one, that we could catch and extend here. I think making `Tensor` throw a more specific exception on shape mismatch, and than catching and re-raising /that one/ here instead makes a lot of sense from the users perspective.

Reply:
But there should never be any exception here. If there is, your code is wrong. If something is not matching, you should have already checked that before.

Reply:
I agree, then the code is wrong. But people are going to realize that their code is wrong by running into the error, at least at first and until they pull out a debugger. And then the error message can help them find the issue.

I changed the code to use a specific exception type, so that only dim mismatches are caught. WDYT about the implementation now?

Reply:
No, I'm not speaking about the user code. I'm speaking about the code here in PostprocessingDataset. There should never be an exception raised here.

Reply:
Ah, but the constructor of Tensor _does_ indeed trigger the shape validation error. Do you think I should create the tensor first, w/o the raw data, and then assign it after the fact in a second statement to make the validation step clearer? Or is there any other good way to validate that the data matches the shape set in the tensor?

Reply:
I mean I can always just extract & re-use the logic from the tensor class to make the validation explicit

Reply:
But PostprocessingDataset should already check it. You cannot rely that some other code does some checks for you.